components/wifi/COMPONENT_QCA400X_WIFI/QCA400XInterface.cpp

```diff
@@ -0,0 +1,676 @@
+    MBED_ASSERT(!_connect_event_id);
+
+    _connect_event_id = _global_event_queue->call(callback(this, &QCA400XInterface::_connect_async));
+
+    if (!_connect_event_id) {
+        MBED_ERROR(MBED_MAKE_ERROR(MBED_MODULE_DRIVER, MBED_ERROR_CODE_ENOMEM), \
+                   "connect(): unable to add event to queue. Increase \"events.shared-eventsize\"\n");
+    }
+
+    while (_if_blocking && (_conn_status_to_error() != NSAPI_ERROR_IS_CONNECTED)
+            && (_connect_retval == NSAPI_ERROR_NO_CONNECTION)) {
+        _if_connected.wait();
+    }
+
+    _cmutex.unlock();
+
+    if (!_if_blocking) {
+        return NSAPI_ERROR_OK;
+    } else {
+        return _connect_retval;
+    }
+}
+
+int QCA400XInterface::set_credentials(const char *ssid, const char *pass,
+                                      nsapi_security_t security)
+{
+    nsapi_error_t status = _conn_status_to_error();
+    if (status != NSAPI_ERROR_NO_CONNECTION) {
+        return NSAPI_ERROR_NO_CONNECTION;
+    }
+
+    _ap_sec = security;
+
+    if (!ssid) {
+        return NSAPI_ERROR_PARAMETER;
+    }
+
+    int ssid_length = strlen(ssid);
+
+    if (ssid_length > 0 && ssid_length <= QCA400X_SSID_MAX_LENGTH) {
+        memset(_ssid, 0, sizeof(_ssid));
+        strncpy(_ssid, ssid, sizeof(_ssid));
+    } else {
+        return NSAPI_ERROR_PARAMETER;
+    }
+
+    if (_ap_sec != NSAPI_SECURITY_NONE) {
+
+        if (!pass) {
+            return NSAPI_ERROR_PARAMETER;
+        }
+
+        int pass_length = strlen(pass);
+        if (pass_length >= QCA400X_PASSPHRASE_MIN_LENGTH
+                && pass_length <= QCA400X_PASSPHRASE_MAX_LENGTH) {
+            memset(ap_pass, 0, sizeof(ap_pass));
+            strncpy(ap_pass, pass, sizeof(ap_pass));
+        } else {
+            return NSAPI_ERROR_PARAMETER;
+        }
+    } else {
+        memset(ap_pass, 0, sizeof(ap_pass));
+    }
+
+    return NSAPI_ERROR_OK;
+}
+
+int QCA400XInterface::set_channel(uint8_t channel)
+{
+    return NSAPI_ERROR_UNSUPPORTED;
+}
+
+int QCA400XInterface::disconnect()
+{
+    _cmutex.lock();
+    if (_connect_event_id) {
+        _global_event_queue->cancel(_connect_event_id);
+        _connect_event_id = 0; // cancel asynchronous connection attempt if one is ongoing
+    }
+    _cmutex.unlock();
+
+    nsapi_error_t status = _conn_status_to_error();
+    if (status == NSAPI_ERROR_NO_CONNECTION) {
+        return NSAPI_ERROR_NO_CONNECTION;
+    }
+
+    int ret = _qca400x.disconnect() ? NSAPI_ERROR_OK : NSAPI_ERROR_DEVICE_ERROR;
+
+    return ret;
+}
+
+const char *QCA400XInterface::get_ip_address()
+{
+    const char *ip_buff = _qca400x.getIPAddress();
+    if (!ip_buff || strcmp(ip_buff, "0.0.0.0") == 0) {
+        return NULL;
+    }
+
+    return ip_buff;
+}
+
+nsapi_error_t QCA400XInterface::get_ip_address(SocketAddress *address)
+{
+    const char *ip_buff = _qca400x.getIPAddress();
+    if (!ip_buff || strcmp(ip_buff, "0.0.0.0") == 0) {
+        return NULL;
```

Comment:
This function should return an error code, rather than `NULL`. I suggest you either do `return NSAPI_ERROR_NO_ADDRESS;` or `ip_buff = NULL;`.